src/extendedstorage.cpp

```diff
@@ -507,8 +507,7 @@ bool ExtendedStorage::deleteNotebook(const Notebook::Ptr &nb)
         const Incidence::List list = calendar()->incidences(nb->uid());
         qCDebug(lcMkcal) << "deleting" << list.size() << "incidences of notebook" << nb->name();
         for (const Incidence::Ptr &toDelete : list) {
-            if (!toDelete->hasRecurrenceId()
-                || calendar()->incidence(toDelete->uid(), toDelete->recurrenceId()))
+            if (calendar()->incidence(toDelete->uid(), toDelete->recurrenceId()))
```

Comment:
Earlier this was deleting events without recurrence id directly from the list of just fetched incidences on a notebook. On what case is it possible that such an incidence doesn't exist anymore?

Reply:
That's a good question. Let me summarize what happened here :
- it was implemented before differently, like
```
Incidence::List list = allIncidences(nb->uid());
for (Incidence::Ptr inc : list) {
  mCalendar->load(inc->uid(), inc->recurrenceId());
  mCalendar->deleteIncidence(inc);
}
```
which was working fine but had the drawback to call `Storage::load()` for every incidence of the calendar, after a call to `allIncidence()`. Because the latter is creating a list of incidences, but not attached to a calendar, while we need them to be in calendar to get them deleted.
- at one point, I thought it would be smart to use instead `loadNotebookIncidences()` because it is loading all incidences of a given notebook in one call to the DB. Then, we can create the list of incidence to delete with a call to `Calendar::incidences(nbuid)`. That was looking great at that moment.
- then, I noticed that `Calendar::deleteIncidence()` on an incidence with exceptions is also deleting the exception (which looks good by itself). But it's an issue if the parent is coming first in the list of incidences, because then, when we want to delete an exception, we're getting a warning from KCalendarCore that the incidence is not in the calendar any more. I circumvent the issue with an ugly `if` where I tested that the incidence was still in the calendar before calling `deleteIncidence()`, but only in the case of an exception (`hasRecurringId() && mCalendar->incidence(uid, recid)`).
- later yesterday, when I added notebook deletion in `tst_load` while there were some deletions already, I stumbled upon a "bug" of KCalendarCore : if you delete some incidences from a calendar, they are still listed by `Calendar::incidences(nbuid)`. So I decided then to simplify my ugly `if` and test existence of any incidence before calling the `deleteIncidence()` function.

I hope my explanation is clear... But let think about this KCalendarCore "bug". I'm not sure it is really a bug or is it by design. I wonder. Actually any deleted incidence is still related to a notebook. I mean, it makes sense to ask, for a deleted incidence to which notebook it was belonging to. And we do this by calling `mCalendar->notebook(incidence)`. So internally the calendar needs to keep the association between incidence and notebook. More over the `Calendar::incidences(nbuid)` method may be interpreted like, "give me all incidence associated to `nbuid`, deleted or not", and not like "give me all incidence associated to `nbuid`, still managed by this calendar". I think the later actually makes more sense, but it could be a major behaviour break to impose it in KCalendarCore.

All in all, we have three options:
- go back to `allIncidences()` way. It was not optimal in term of DB access, but it was full proof.
- live with the KCalendarCore "bug" or behaviour and accept the ugly `mCalendar->incidence()` test before the `deleteIncidence()` call.
- try to convince KCalendarCore to change `Calendar::incidences(nbuid)` not to return deleted incidences, but we will still have the ugly `if` because any exceptions may have been deleted already with its parent.

Frankly, I hesitated with the first option, but the second looks slightly better, limiting DB calls.

Reply:
Woah, that's really something. Yea, looks like the "earlier" was before commit a9b7aed63a47 which was effectively the same in the initial commit here.

Fair enough, deleting a notebook is a rare operation so the extra check shouldn't matter much and it's a simple way to avoid problems. But perhaps could add some note here why it's done.

Reply:
I've added a somehow lengthy comment summarising the above discussion near the `if` test to explain its use.